test/test_common/simulated_time_system.h

```diff
@@ -0,0 +1,92 @@
+#pragma once
+
+#include "envoy/event/timer.h"
+
+#include "common/common/lock_guard.h"
+#include "common/common/thread.h"
+#include "common/common/utility.h"
+
+namespace Envoy {
+namespace Event {
+
+// Represents a simulated time system, where time is advanced by calling
+// sleep(), setSystemTime(), or setMonotonicTime(). systemTime() and
+// monotonicTime() are maintained in the class, and alarms are fired in response
+// to adjustments in time.
+class SimulatedTimeSystem : public TimeSystem {
+public:
+  SimulatedTimeSystem();
+
+  // TimeSystem
+  SchedulerPtr createScheduler(Libevent::BasePtr&) override;
+
+  // TimeSource
+  SystemTime systemTime() override;
+  MonotonicTime monotonicTime() override;
+
+  /**
+   * Advances time forward by the specified duration, running any timers
+   * along the way that have been scheduled to fire.
+   *
+   * @param duration The amount of time to sleep, expressed in any type that
+   * can be duration_casted to MonotonicTime::duration.
+   */
+  template <class Duration> void sleep(const Duration& duration) {
+    mutex_.lock();
+    MonotonicTime monotonic_time =
+        monotonic_time_ + std::chrono::duration_cast<MonotonicTime::duration>(duration);
+    setMonotonicTimeAndUnlock(monotonic_time);
+  }
+
+  /**
+   * Sets the time forward monotonically. If the supplied argument moves
+   * backward in time, the call is a no-op. If the supplied argument moves
+   * forward, any applicable timers are fired, and system-time is also moved
+   * forward by the same delta.
+   *
+   * @param monotonic_time The desired new current time.
+   */
+  void setMonotonicTime(const MonotonicTime& monotonic_time) {
+    mutex_.lock();
+    setMonotonicTimeAndUnlock(monotonic_time);
+  }
+
+  /**
+   * Sets the system-time, whether forward or backward. If time moves forward,
+   * applicable timers are fired and monotonic time is also increased by the
+   * same delta.
+   *
+   * @param system_time The desired new system time.
+   */
+  void setSystemTime(const SystemTime& system_time);
+
+private:
+  class SimulatedScheduler;
+  friend class SimulatedScheduler;
+  class Alarm;
```

Comment:
Not sure about whether it is idiomatic...I have a practice of trying to put impl details into the cc where possible and not impactful to performance.

IMO that helps hide unimportant details from the user of the simulated time system, keep the impl details together, and minimize recompiles when changing impl.

But i can put them here if it is a strong envoy convention.

Reply:
That seems to me like a fair assessment. Good to know the intention, and maybe something I can adopt in the future. 

I mentioned it because in general what I have seen in the envoy codebase is the other way around, where, if small enough, implementation is brought into the header file. But I don't think there is a hard and fast rule.

@alyssawilk or @htuch just wondering if you lean in one way or another in this topic?

Reply:
btw based on past experience, I think Envoy coding style of inling virtual dtors may contribute to slow link-time for test binaries, which we have a lot of.

In Chrome coding style this inlining of all ctors/dtors is discouraged, though IMO the virtual dtors are more of an issue.

Reply:
The other consideration is rebuild time; if you have to make a change to a `.h` that is imported in 100 places you have 100 rebuilds, vs. the single `.cc`. I agree with @junr03 that we have a habit of putting small implementations in `.h`. I think this is because (i) it's less typing and (ii) more readable. I think link and rebuild time is a good argument for preferring `.cc`, but no strong preference.

Reply:
Do we really need to do as much forward declaration here? Is this to break a cycle?

Reply:
No, it's to declare friend classes that only need to be ref'd internal to the impl.